### PR TITLE
[PPWM-81] Limit number of workspace migration retries

### PR DESCRIPTION
**Pull request by ehigham**
RR: https://broadworkbench.atlassian.net/browse/PPWM-81
Changes:
- Use a new table to record the number fo times a migration has been retried. Don't keep retrying when the number exceeds some configured value.
- Ensure that the workspace is unlocked if a previous migration locked the workspace.

- [ ] **Submitter**: Include the JIRA issue number in the PR description
- [ ] **Submitter**: Check that the **Product Owner** has signed off on any user-facing changes
- [ ] **Submitter**: Make sure Swagger is updated if API changes
  - [ ] **...and Orchestration's Swagger too!**
- [ ] **Submitter**: If updating admin endpoints, also update [firecloud-admin-cli](https://github.com/broadinstitute/firecloud-admin-cli)
- [ ] **Submitter**: Check documentation and code comments. Add explanatory PR comments if helpful.
- [ ] **Submitter**: JIRA ticket checks:
  * Acceptance criteria exists and is met
  * Note any changes to implementation from the description
  * To Demo flag is set
  * Release Summary is filled out, if applicable
  * Add notes on how to QA
- [ ] **Submitter**: Update RC_XXX release ticket with any config or environment changes necessary
- [ ] **Submitter**: Database checks:
  * If PR includes new or changed db queries, include the explain plans in the description
  * Make sure liquibase is updated if appropriate
  * If doing a migration, take a backup of the
  [dev](https://console.cloud.google.com/sql/instances/terraform-qfarbdq3lrexxck5htofjs5z6m/backups?project=broad-dsde-dev&organizationId=548622027621)
  and
  [alpha](https://console.cloud.google.com/sql/instances/terraform-r4caezzc35c4tb7pgdhwkmme4y/backups?project=broad-dsde-alpha&organizationId=548622027621)
  DBs in Google Cloud Console
- [ ] **Submitter**: Update FISMA documentation if changes to:
  * Authentication
  * Authorization
  * Encryption
  * Audit trails
- [ ] Tell your tech lead (TL) that the PR exists if they want to look at it
- [ ] Anoint a lead reviewer (LR). **Assign PR to LR**
* Review cycle:
  * LR reviews
  * Rest of team may comment on PR at will
  * **LR assigns to submitter** for feedback fixes
  * Submitter rebases to develop again if necessary
  * Submitter makes further commits. DO NOT SQUASH
  * Submitter updates documentation as needed
  * Submitter **reassigns to LR** for further feedback
- [ ] **TL** sign off
- [ ] **LR** sign off
- [ ] **Assign to submitter** to finalize
- [ ] **Submitter**: Verify all tests go green, including CI tests
- [ ] **Submitter**: Squash commits and merge to develop
- [ ] **Submitter**: Delete branch after merge
- [ ] **Submitter**: **Test this change works on dev environment after deployment**. YOU own getting it fixed if dev isn't working for ANY reason!
- [ ] **Submitter**: Verify swagger UI on dev environment still works after deployment
- [ ] **Submitter**: Inform other teams of any API changes via Slack and/or email
- [ ] **Submitter**: Mark JIRA issue as resolved once this checklist is completed

### CODE REVIEW

File: core/src/main/scala/org/broadinstitute/dsde/rawls/monitor/migration/WorkspaceMigration.scala [hunk 1]
```diff
@@ -368,8 +301,131 @@ trait WorkspaceMigrationHistory extends DriverComponent with RawSqlQuery {
...
+    val noPermissionsFailure: String =
+      "%FAILED_PRECONDITION: Service account project-%@storage-transfer-service.iam.gserviceaccount.com " +
+        "does not have required permissions%"
+
+    val rateLimitedFailure: String =
+      "%RESOURCE_EXHAUSTED: Quota exceeded for quota metric 'Create requests' " +
+        "and limit 'Create requests per day' of service 'storagetransfer.googleapis.com'%"
+  }
+
+  case class RawTableQuery[PrimaryKey](tableName: String, primaryKey: ColumnName[PrimaryKey])
```
Comment (ehigham) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/monitor/migration/WorkspaceMigration.scala:345]: This class is used to share the `update` functionality across the various tables

File: core/src/main/scala/org/broadinstitute/dsde/rawls/monitor/migration/WorkspaceMigration.scala [hunk 2]
```diff
@@ -368,8 +301,131 @@ trait WorkspaceMigrationHistory extends DriverComponent with RawSqlQuery {
...
+
 
+    def delete: WriteAction[Int] =
+      sqlu"""delete from #$tableName"""
   }
 }
 
-final case class MigrationColumnName(value: String) extends ValueObject
\ No newline at end of file
+final case class ColumnName[-A](value: String) extends ValueObject
```
Comment (ehigham) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/monitor/migration/WorkspaceMigration.scala:422]: Using a type parameter helps ensure that the type of the column matches the value in `update`

File: core/src/main/scala/org/broadinstitute/dsde/rawls/monitor/migration/WorkspaceMigrationActor.scala [hunk 3]
```diff
@@ -1124,8 +1140,11 @@ object WorkspaceMigrationActor {
     }
 
 
-  final def inTransactionT[A](action: DataAccess => ReadWriteAction[Option[A]]): MigrateAction[A] =
-    inTransaction(action).mapF(optT => OptionT(optT.value.map(_.flatten)))
+  final def inTransactionT[A](action: DataAccess => OptionT[ReadWriteAction, A]): MigrateAction[A] =
```
Comment (ehigham) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/monitor/migration/WorkspaceMigrationActor.scala:1143]: Most of the changes in this file are becuase of this line - this allows me to run a transaction over `OptionT` which is quite handy for guarding execution and dealing with optionals without wrapping the for-comprehension.

File: core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/WorkspaceComponent.scala [hunk 4]
```diff
@@ -282,14 +282,6 @@ trait WorkspaceComponent {
...
     }
 
-    def lock(workspaceName: WorkspaceName): ReadWriteAction[Boolean] = {
-      findByNameQuery(workspaceName).filter(!_.isLocked).map(_.isLocked).update(true).map(_ > 0)
-    }
-
-    def unlock(workspaceName: WorkspaceName): ReadWriteAction[Int] = {
-      findByNameQuery(workspaceName).map(_.isLocked).update(false)
-    }
-
```
Comment (ehigham) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/WorkspaceComponent.scala:292]: All uses of these had the workspace id already

File: core/src/main/scala/org/broadinstitute/dsde/rawls/monitor/migration/WorkspaceMigrationActor.scala [hunk 5]
```diff
@@ -749,42 +763,39 @@ object WorkspaceMigrationActor {
     }
 
 
-  def restartTransferServiceNoPermissions: MigrateAction[Unit] =
+  def retry: MigrateAction[Unit] =
```
Comment (ehigham) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/monitor/migration/WorkspaceMigrationActor.scala:766]: This function is the main change

**Review by dvoet — APPROVED** [review]

**Comment by ehigham** [comment]
jenkins retest

**Pull request merged**